### ISSUE TITLE
[Bug] Error when decreasing quantity of order item in cart

### ISSUE THREAD

**Issue by steffenbrem** [Sylius/Sylius#4005]
There is currently a big bug in Sylius with the `OrderItemsTaxesByZoneApplicator`. When you have a Book in your cart with a quantity of `5` and then change it to `4` and then save your cart, it gives this error:

```
Type error: Argument 1 passed to
Sylius\Component\Core\Taxation\OrderItemsTaxesByZoneApplicator::addAdjustment() 
must implement interface Sylius\Component\Core\Model\OrderItemUnitInterface, null given, 
called in /var/www/html/vendor/sylius/sylius/src/Sylius/Component/Core/Taxation/OrderItemsTaxesByZoneApplicator.php 
on line 88
```

Tested it on latest master (8b479af8a20f06ecd1121fa11f59f1d4246da810) and the bug is also present on http://demo.sylius.org/

@Zales0123 Do you know whatsup?

As you can see on this screenshot, the array keys are different, one starts with `0` and the `$units` starts with `1`. I guess something is going wrong there?
![Debug](http://puu.sh/mMXdW/2e334f1517.png)

**Comment by steffenbrem** [comment]
Would resetting the keys of `$units` be a good solution?

**Comment by michalmarcinkowski** [comment]
Yeah I thought there can be issue with [this](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Component/Core/Taxation/OrderItemsTaxesByZoneApplicator.php#L88-L88) implementation. Probably the element that is removed from the array has index `0` and then we call `$unitsCollection[0]` which returns `null`. Will fix this ASAP.

@steffenbrem thanks for reporting! :+1:

**Comment by steffenbrem** [comment]
@michalmarcinkowski I am currently using this to solve it for me for now:

```php
$units = $item->getUnits()->getValues();
foreach ($splitTaxes as $key => $tax) {
    if (0 === $tax) {
        continue;
    }

    $this->addAdjustment($units[$key], $tax, $taxRate->getLabel(), $taxRate->isIncludedInPrice());
}
```

**Comment by michalmarcinkowski** [comment]
I would go with safer solution:
```php
$units = $item->getUnits();
foreach ($splitTaxes as $key => $tax) {
        if (0 === $tax) {
              continue;
        }

        $unit = $units->current();
        if (null === $unit) {
            throw new \InvalidArgumentException('The number of tax items is greater than number of units.');
        }

        $this->addAdjustment($unit, $tax, $taxRate->getLabel(), $taxRate->isIncludedInPrice());
}
```
Will prepare a PR today.

**Comment by pjedrzejewski** [comment]
Fixed in #4012, Thanks @steffenbrem for reporting! :+1:
